Source/GSOrderedSet.m

```diff
@@ -0,0 +1,358 @@
+@end
+
+@implementation GSOrderedSetEnumerator
+- (id) initWithOrderedSet: (NSOrderedSet*)d
+{
+  self = [super init];
+  if (self != nil)
+    {
+      set = (GSOrderedSet*)RETAIN(d);
+      current = 0;
+      count = GSIArrayCount(&set->array);
+    }
+  return self;
+}
+
+- (id) nextObject
+{
+  if(current < count)
+    {
+      GSIArrayItem item = GSIArrayItemAtIndex(&set->array, current);
+      current++;
+      return (id)(item.obj);
+    }
+  return nil;
+}
+
+- (void) dealloc
+{
+  RELEASE(set);
+  [super dealloc];
+}
+@end
+
+
+@implementation GSOrderedSetEnumeratorReverse
+- (id) initWithOrderedSet: (GSOrderedSet*)d
+{
+  self = [super initWithOrderedSet: d];
+  if(self != nil)
+    {
+      current = GSIArrayCount(&set->array);
+    }
+  return self;
+}
+
+- (id) nextObject
+{
+  GSIArrayItem item;
+
+  if (current == 0)
+    return nil;
+  
+  item = GSIArrayItemAtIndex(&set->array, --current);
+  return (id)(item.obj);
+}
+@end
+
+@implementation GSOrderedSet
+
+static Class	setClass;
+static Class	mutableSetClass;
+
++ (void) initialize
+{
+  if (self == [GSOrderedSet class])
+    {
+      setClass = [GSOrderedSet class];
+      mutableSetClass = [GSMutableOrderedSet class];
+    }
+}
+
+- (id) copyWithZone: (NSZone*)z
+{
+  return RETAIN(self);
+}
+
+- (void) dealloc
+{
+  GSIArrayEmpty(&array);
+  [super dealloc];
+}
+
+- (NSUInteger) hash
+{
+  return [self count];
+}
+
+- (instancetype) init
+{
+  return [self initWithObjects: NULL count: 0];
+}
+
+- (NSEnumerator*) objectEnumerator
+{
+  return AUTORELEASE([[GSOrderedSetEnumerator alloc] initWithOrderedSet: self]);
+}
+
+- (NSEnumerator*) reverseObjectEnumerator
+{
+  return AUTORELEASE([[GSOrderedSetEnumeratorReverse alloc] initWithOrderedSet: self]);
+}
+
+- (NSUInteger) sizeInBytesExcluding: (NSHashTable*)exclude
+{
+  NSUInteger	size = GSPrivateMemorySize(self, exclude);
+
+  if (size > 0)
+    {
+      NSUInteger count = [self count];
+      NSUInteger i = 0;
+
+      for(i = 0; i < count; i++)
+	{
+	  GSIArrayItem item = GSIArrayItemAtIndex(&array, i);
+          size += [item.obj sizeInBytesExcluding: exclude];
+        }
+    }
+  return size;
+}
+
+// Put required overrides here...
+- (NSUInteger) count
+{
+  return GSIArrayCount(&array);
+}
+
+- (id) objectAtIndex: (NSUInteger)index
+{
+  GSIArrayItem item = GSIArrayItemAtIndex(&array, index);
+  return item.obj;
+}
+
+/* Designated initialiser */
+- (id) initWithObjects: (const id*)objs count: (NSUInteger)c
+{
+  NSUInteger i = 0;
+
+  // Initialize and fill the set.
+  GSIArrayInitWithZoneAndCapacity(&array, [self zone], c);
+  for (i = 0; i < c; i++)
+    {
+      id obj = objs[i];
+      GSIArrayItem item;
+      
+      if (objs[i] == nil)
+	{
+	  DESTROY(self);
+	  [NSException raise: NSInvalidArgumentException
+		      format: @"Tried to init set with nil value"];
+	}
+
+      item.obj = obj;
+      if(![self containsObject: obj])
+	{
+	  GSIArrayAddItem(&array, item);
+	  RETAIN(obj);
```

Comment:
If the GSIArray already retains the object, we don't need to do it here (and in all the GSMutableOrderedSet methods).

Reply:
Seems to crash unless I do.  I have GSIArray configured to retain, but I'm not sure why it's losing it here.

Reply:
This bit is important. We should not just ignore it. I have some time today and will try to look into the issue.

Reply:
Please do... I want to close this out and get it merged.

Reply:
It looks like you addressed this.   I would very much like to merge this at this point.